Google.PowerShell/Storage/GcsObject.cs

```diff
@@ -922,4 +917,75 @@ protected override void ProcessRecord()
             }
         }
     }
+
+    [Cmdlet(VerbsCommon.Copy, "GcsObject", DefaultParameterSetName = ParameterSetNames.ByObject)]
+    [OutputType(typeof(Object))]
+    public class CopyGcsObject : GcsCmdlet
+    {
+
+        private class ParameterSetNames
+        {
+            public const string ByName = "ByName";
+            public const string ByObject = "ByObject";
+        }
+
+        [Parameter(ParameterSetName = ParameterSetNames.ByObject, Mandatory = true, ValueFromPipeline = true)]
+        public Object InputObject { get; set; }
+
+        /// <summary>
+        /// <para type="description">
+        /// Name of the bucket containing the object. Will also accept a Bucket object.
+        /// </para>
+        /// </summary>
+        [Parameter(ParameterSetName = ParameterSetNames.ByName, Mandatory = true)]
+        [PropertyByTypeTransformation(Property = nameof(Bucket.Name), TypeToTransform = typeof(Bucket))]
+        public string SourceBucket { get; set; }
+
+        /// <summary>
+        /// <para type="description">
+        /// Name of the object to write to.
+        /// </para>
+        /// </summary>
+        [Parameter(ParameterSetName = ParameterSetNames.ByName, Mandatory = true)]
+        public string SourceObjectName { get; set; }
+
+
+        /// <summary>
+        /// <para type="description">
+        /// Name of the bucket in which the copy will reside.
+        /// </para>
+        /// </summary>
+        [Parameter(Mandatory = false, Position = 0)]
+        [PropertyByTypeTransformation(Property = nameof(Bucket.Name), TypeToTransform = typeof(Bucket))]
+        public string DestinationBucket { get; set; }
+
+        /// <summary>
+        /// <para type="description">
+        /// The name of the copy.
+        /// </para>
+        /// </summary>
+        [Parameter(Mandatory = false, Position = 1)]
+        public string DestinationObjectName { get; set; }
+
+        protected override void ProcessRecord()
+        {
+            Object gcsObject;
+            switch (ParameterSetName)
+            {
+                case ParameterSetNames.ByName:
+                    gcsObject = Service.Objects.Get(SourceBucket, SourceObjectName).Execute();
+                    break;
+                case ParameterSetNames.ByObject:
+                    gcsObject = InputObject;
+                    break;
+                default:
+                    throw UnknownParameterSetException;
+            }
+            ObjectsResource.CopyRequest request = Service.Objects.Copy(gcsObject,
+                gcsObject.Bucket, gcsObject.Name,
+                DestinationBucket ?? gcsObject.Bucket, DestinationObjectName ?? gcsObject.Name);
```

Comment:
Having defaults for these values seems reasonable. But you should mention that in the doc comments above.